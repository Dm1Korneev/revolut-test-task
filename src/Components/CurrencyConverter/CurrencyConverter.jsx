import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { RATES_RELOAD_PERIOD } from 'Constants';

import PocketFrom from 'Components/PocketFrom';
import PocketTo from 'Components/PocketTo';
import CurrentRate from 'Components/CurrentRate';
import ExchangeButton from 'Components/ExchangeButton';

import { Title } from './CurrencyConverter.styled';

const propTypes = {
  getPockets: PropTypes.func.isRequired,
  getRates: PropTypes.func.isRequired,
};

class CurrencyConverter extends Component {
  constructor(props) {
    super(props);

    this.interval = undefined;
  }

  componentDidMount() {
    const { getRates, getPockets } = this.props;

    getPockets();
    getRates();
    this.interval = setInterval(getRates, RATES_RELOAD_PERIOD);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render() {
    return (
      <>
        <Title>Currency exchange</Title>
        <CurrentRate />
        <ExchangeButton />
        <PocketFrom />
        <PocketTo />
      </>
    );
  }
}

CurrencyConverter.propTypes = propTypes;

export default CurrencyConverter;
