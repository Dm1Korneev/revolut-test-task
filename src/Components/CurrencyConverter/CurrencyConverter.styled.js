import styled from 'styled-components';

const Currency = styled.div`
  width: 400px;
  height: 200px;
`;

export const CurrencyFrom = styled(Currency)`
  background-color: red;
`;

export const CurrencyTo = styled(Currency)`
  background-color: blue;
`;

export const Title = styled.h1`
  font-family: "Futura New", Futura, Avenir, sans-serif;
  font-weight: 400;
  line-height: 1.25;
  font-size: 3rem;
  text-align: center;
  padding-bottom: 1rem;
`;
