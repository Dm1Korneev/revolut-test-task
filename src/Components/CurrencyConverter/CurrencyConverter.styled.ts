import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 30rem;
  border: 1px solid rgba(0, 0, 0, 0.12);
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2), 
              0px 1px 1px 0px rgba(0,0,0,0.14), 
              0px 1px 3px 0px rgba(0,0,0,0.12);
`;

export const BlockFrom = styled.div`
display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1rem;
  background-color: rgb(255, 255, 255);
  position: relative;
`;

export const BlockTo = styled.div`
display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1rem;
  border-top: 1px solid rgba(0, 0, 0, 0.12);
  background-color: rgb(243, 243, 243);
`;

export const CurrentRateContainer = styled.div`
    position: absolute;
    bottom: -1rem;
    height: 2rem;
    display: flex;
    align-items: center;
    background: rgb(255, 255, 255);
    padding: 1rem;
    border-radius: 1rem;
    border: 1px solid rgba(0,0,0,0.12);
`;

export const ChangePocketsContainer = styled.div`
    position: absolute;
    left: 1rem;
    bottom: -1rem;
    height: 2rem;
    width: 2rem;
    display: flex;
    align-items: center;
    background: rgb(255, 255, 255);
    border-radius: 1rem;
    border: 1px solid rgba(0,0,0,0.12);
`;

export const Title = styled.h1`
  font-family: "Futura New", Futura, Avenir, sans-serif;
  font-weight: 400;
  line-height: 1.25;
  font-size: 3rem;
  text-align: center;
  padding: 1rem;
`;
