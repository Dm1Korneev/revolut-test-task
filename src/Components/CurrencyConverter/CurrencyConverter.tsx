import React, { Component } from 'react';

import { RATES_RELOAD_PERIOD } from 'Constants';

import PocketFrom from 'Components/PocketFrom';
import PocketTo from 'Components/PocketTo';
import CurrentRate from 'Components/CurrentRate';
import ExchangeButton from 'Components/ExchangeButton';
import Loader from 'Components/Loader';
import ChangePocketsButton from 'Components/ChangePocketsButton';

import {
  BlockFrom, BlockTo, ChangePocketsContainer, Container, CurrentRateContainer, Title,
} from './CurrencyConverter.styled';

type Props = {
  getRates: Function;
  getPockets: Function;
  isBusy: boolean;
}

class CurrencyConverter extends Component<Props> {
  interval: number;

  constructor(props) {
    super(props);

    this.interval = undefined;
  }

  componentDidMount(): void {
    const { getRates, getPockets } = this.props;

    getPockets();
    getRates();
    this.interval = setInterval(getRates, RATES_RELOAD_PERIOD);
  }

  componentWillUnmount(): void {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render(): JSX.Element {
    const { isBusy } = this.props;

    return (
      <Container>
        <Title>Currency exchange</Title>
        {isBusy ? <Loader /> : (
          <>
            <BlockFrom>
              <PocketFrom />
              <CurrentRateContainer>
                <CurrentRate />
              </CurrentRateContainer>
              <ChangePocketsContainer>
                <ChangePocketsButton />
              </ChangePocketsContainer>
            </BlockFrom>
            <BlockTo>
              <PocketTo />
              <ExchangeButton />
            </BlockTo>
          </>
        )}
      </Container>
    );
  }
}

export default CurrencyConverter;
