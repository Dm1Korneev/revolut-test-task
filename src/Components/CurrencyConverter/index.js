import commonHoc from 'Components/commonHoc';
import { getPockets, getRates } from 'Redux/actions';
import { getRatesLoadingSelector } from 'Selectors/loading';

import CurrencyConverter from './CurrencyConverter';

const mapStateToProps = (state) => ({
  isBusy: getRatesLoadingSelector(state),
});

const mapDispatchToProps = {
  getRates,
  getPockets,
};

export default commonHoc(CurrencyConverter, {
  mapStateToProps,
  mapDispatchToProps,
});
