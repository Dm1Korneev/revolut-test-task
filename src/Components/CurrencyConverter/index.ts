import commonHoc from 'Components/commonHoc';
import { getPockets, getRates } from 'Redux/actions';
import { getRatesLoadedSelector } from 'Selectors/loaded';
import { StateType } from 'Redux/store';

import CurrencyConverter from './CurrencyConverter';

const mapStateToProps = (state: StateType): Record<string, any> => ({
  isBusy: !getRatesLoadedSelector(state),
});

const mapDispatchToProps = {
  getRates,
  getPockets,
};

export default commonHoc(CurrencyConverter, {
  mapStateToProps,
  mapDispatchToProps,
});
