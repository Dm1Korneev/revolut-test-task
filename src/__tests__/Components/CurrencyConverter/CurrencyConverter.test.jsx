import React from 'react';
import { render } from '@testing-library/react';
import CurrencyConverter from 'Components/CurrencyConverter/CurrencyConverter';

jest.mock('Components/PocketFrom', () => global.mockComponent('PocketFrom'));
jest.mock('Components/PocketTo', () => global.mockComponent('PocketTo'));
jest.mock('Components/CurrentRate', () => global.mockComponent('CurrentRate'));
jest.mock('Components/ExchangeButton', () => global.mockComponent('ExchangeButton'));
jest.mock('Components/Loader', () => global.mockComponent('Loader'));
jest.mock('Components/ChangePocketsButton', () => global.mockComponent('ChangePocketsButton'));

describe('CurrencyConverter component', () => {
  const props = {
    getPockets: jest.fn(),
    getRates: jest.fn(),
    isBusy: false,
  };

  test('renders the component', () => {
    const { asFragment } = render(<CurrencyConverter {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('show loader when isBusy = true', () => {
    const propsLocal = { ...props, isBusy: true };
    const { asFragment } = render(<CurrencyConverter {...propsLocal} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('call get methods', () => {
    const getPockets = jest.fn();
    const getRates = jest.fn();
    const propsLocal = { ...props, getPockets, getRates };

    render(<CurrencyConverter {...propsLocal} />);
    expect(getPockets).toHaveBeenCalledTimes(1);
    expect(getRates).toHaveBeenCalledTimes(1);
  });
});
