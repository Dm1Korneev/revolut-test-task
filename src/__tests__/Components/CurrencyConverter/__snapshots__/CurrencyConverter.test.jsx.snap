// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter component renders the component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa dBnMMB"
  >
    <h1
      class="sc-bZQynM cvwvqF"
    >
      Currency exchange
    </h1>
    <div
      class="sc-bwzfXH sc-htpNat dtmtwx"
    >
      <div
        originalcomponent="PocketFrom"
      />
      <div
        class="sc-ifAKCX gHNnsm"
      >
        <div
          originalcomponent="CurrentRate"
        />
      </div>
      <div
        class="sc-EHOje btOkpo"
      >
        <div
          originalcomponent="ChangePocketsButton"
        />
      </div>
    </div>
    <div
      class="sc-bwzfXH sc-bxivhb eMThae"
    >
      <div
        originalcomponent="PocketTo"
      />
      <div
        originalcomponent="ExchangeButton"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`CurrencyConverter component show loader when isBusy = true 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa dBnMMB"
  >
    <h1
      class="sc-bZQynM cvwvqF"
    >
      Currency exchange
    </h1>
    <div
      originalcomponent="Loader"
    />
  </div>
</DocumentFragment>
`;
